### PR TITLE
Add an internal solution for CF 472D

### PR DESCRIPTION
**Pull request by chessbot108**
The Cf editorial did not come with code.

### CODE REVIEW

File: solutions/cf-472D.mdx [hunk 1]
```diff
@@ -0,0 +1,121 @@
+---
+id: cf-472D
+source: Codeforces Round #270 (472)
```
Comment (dongliu0426) [inline solutions/cf-472D.mdx:3]:
```suggestion
source: Codeforces
```

File: solutions/cf-472D.mdx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+title: Design Tutorial Inverse the Problem
+author: Weiming Zhou
+---
+
+[Official Analysis](https://codeforces.com/blog/entry/14028)
+
+**Time Complexity:** $\mathcal{O}(N^2\ logN)$ with Kruskal's MST or $\mathcal{O}(N^2)$ with N^2 Prims.
```
Comment (dongliu0426) [inline solutions/cf-472D.mdx:10]:
```suggestion
**Time Complexity:** $\mathcal{O}(N^2\log N)$ with Kruskal's MST or $\mathcal{O}(N^2)$ with $N^2$ Prims.
```

File: solutions/cf-472D.mdx [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+int main(){
+	cin.tie(0) -> sync_with_stdio(0);
+	bool ans = 1; 
+	cin >> n;
+	for(int i = 1; i<=n; i++){
+	for(int j = 1; j<=n; j++){
+		cin >> dist[i][j];
+			if(i != j){
+				if(j < i && dist[i][j] != dist[j][i]) ans = 0; 
+			if(i < j) edges.pb(mp(dist[i][j], mp(i, j)));
```
Comment (dongliu0426) [inline solutions/cf-472D.mdx:100]:
```suggestion
				if(i < j) edges.pb(mp(dist[i][j], mp(i, j)));
```
oops

**Review by dongliu0426 — CHANGES_REQUESTED** [review]

**Review by dongliu0426 — APPROVED** [review]

**Comment by sathu-beep** [comment]
code looks sus 
The indenting currently is composed of spaces. you should switch them out for tabs using https://usaco.guide/editor/ 
use right click > command pallette > convert spaces to tabs
other than that looks and works good :D

**Comment by dongliu0426** [comment]
> The indenting currently is composed of spaces.

?

File: solutions/cf-472D.mdx [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+void Union(int a, int b){
+	a = find(a), b = find(b);
+	if(a == b) return ;
+	if(sz[a] > sz[b]) swap(a, b);
+	par[b] = a;
+	sz[a] += sz[b];
+}
+
+void mkdsu(){
+	for(int i = 0; i<=n; i++){
```
Comment (maggie-j-liu) [inline solutions/cf-472D.mdx:54]:
```suggestion
	for(int i = 0; i <= n; i++){
```

File: solutions/cf-472D.mdx [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	for(int j = 1; j<=n; j++){
+		cin >> dist[i][j];
+			if(i != j){
+				if(j < i && dist[i][j] != dist[j][i]) ans = 0; 
+				if(i < j) edges.pb(mp(dist[i][j], mp(i, j)));
+			}else{
+				if(dist[i][j] != 0) ans = 0; 
+			}
+		}
+	}
```
Comment (maggie-j-liu) [inline solutions/cf-472D.mdx:105]:
```suggestion
	for(int i = 1; i <= n; i++){
		for(int j = 1; j <= n; j++){
			cin >> dist[i][j];
			if(i != j){
				if(j < i && dist[i][j] != dist[j][i]) ans = 0; 
				if(i < j) edges.pb(mp(dist[i][j], mp(i, j)));
			}else{
				if(dist[i][j] != 0) ans = 0; 
			}
		}
	}
```

File: solutions/cf-472D.mdx [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+	kruskals();
+
+	for(int i = 1; i<=n; i++){
```
Comment (maggie-j-liu) [inline solutions/cf-472D.mdx:109]:
```suggestion
	for(int i = 1; i <= n; i++){
```

File: solutions/cf-472D.mdx [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+#define pb push_back
+#define all(x) begin(x), end(x)
+#define sz(x) (int)(x).size()
+
+using pii = pair<int,int>; //note pii, not pi
+#define f first
+#define s second
+#define mp make_pair
+
+const int MX = 2e3 +10;
```
Comment (maggie-j-liu) [inline solutions/cf-472D.mdx:34]:
```suggestion
const int MX = 2e3 + 10;
```

File: solutions/cf-472D.mdx [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+vector<pair<int, pii>> edges;
+
+int find(int x){
+	if(x != par[x]) par[x] = find(par[x]); 
+	return par[x];
+}
+
+void Union(int a, int b){
+	a = find(a), b = find(b);
+	if(a == b) return ;
```
Comment (maggie-j-liu) [inline solutions/cf-472D.mdx:47]:
```suggestion
	if(a == b) return;
```

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]
indentation looks fine to me...

File: solutions/cf-472D.mdx [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+	return par[x];
+}
+
+void Union(int a, int b){
+	a = find(a), b = find(b);
+<<<<<<< HEAD
+	if(a == b) return ;
+=======
+	if(a == b) return;
+>>>>>>> 4235be1a4f3175cebfb05c752a7af875255688a9
```
Comment (maggie-j-liu) [inline solutions/cf-472D.mdx:51]: i think something went wrong when merging... the indentation was fine before so you can just revert your latest commit?

Comment (chessbot108) [reply solutions/cf-472D.mdx:52]: uh i think i killed git by being dumb. i will delete the branch and try again

**Pull request closed without merging**